phd-tests/framework/src/disk/crucible.rs

```diff
@@ -0,0 +1,275 @@
+    process_handle: std::process::Child,
+    address: SocketAddr,
+    data_dir: DataDirectory,
+}
+
+impl Drop for Downstairs {
+    fn drop(&mut self) {
+        info!(?self, "Stopping Crucible downstairs process");
+        let _ = self.process_handle.kill();
+    }
+}
+
+/// An RAII wrapper around a Crucible disk.
+#[derive(Debug)]
+pub struct CrucibleDisk {
+    /// The UUID to insert into this disk's `VolumeConstructionRequest`s.
+    id: Uuid,
+
+    /// The disk's block size.
+    block_size: BlockSize,
+
+    /// The collection of downstairs process wrappers for this disk.
+    downstairs_instances: Vec<Downstairs>,
+
+    /// An optional path to a file to use as a read-only parent for this disk.
+    read_only_parent: Option<PathBuf>,
+
+    /// The kind of guest OS that can be found on this disk, if there is one.
+    guest_os: Option<GuestOsKind>,
+
+    /// The generation number to insert into this disk's
+    /// `VolumeConstructionRequest`s.
+    generation: AtomicU64,
+}
+
+impl CrucibleDisk {
+    /// Constructs a new Crucible disk that stores its files in the supplied
+    /// `data_dir`.
+    pub(crate) fn new(
+        disk_size_gib: u64,
+        block_size: BlockSize,
+        downstairs_binary_path: &impl AsRef<std::ffi::OsStr>,
+        downstairs_ports: &[u16],
+        data_dir: &impl AsRef<Path>,
+        read_only_parent: Option<&impl AsRef<Path>>,
+        guest_os: Option<GuestOsKind>,
+        log_mode: ServerLogMode,
+    ) -> anyhow::Result<Self> {
+        // To create a region, Crucible requires a block size, an extent size
+        // given as a number of blocks, and an extent count. Compute the latter
+        // two figures here. The extent size is semi-arbitrarily chosen to be 64
+        // MiB (to match Omicron's extent size at the time this module was
+        // authored); this can be parameterized later if needed.
+        const EXTENT_SIZE: u64 = 64 << 20;
+
+        assert!(EXTENT_SIZE > block_size.bytes());
+        assert!(EXTENT_SIZE % block_size.bytes() == 0);
+
+        let blocks_per_extent: u64 = EXTENT_SIZE / block_size.bytes();
+        let extents_in_disk = (disk_size_gib * (1 << 30)) / EXTENT_SIZE;
+
+        // Create the region directories for each region.
+        let mut data_dirs = vec![];
+        let disk_uuid = Uuid::new_v4();
+        for port in downstairs_ports {
+            let mut data_dir_path = data_dir.as_ref().to_path_buf();
+            data_dir_path.push(format!("{}_{}", disk_uuid, port));
+            std::fs::create_dir_all(&data_dir_path)?;
+            data_dirs.push(DataDirectory { path: data_dir_path });
+        }
+
+        for dir in &data_dirs {
+            let dir_arg = dir.path.to_string_lossy();
+            let crucible_args = [
+                "create",
+                "--block-size",
+                &block_size.bytes().to_string(),
+                "--data",
+                dir_arg.as_ref(),
+                "--encrypted",
+                "false",
```

Comment:
It's false by default, but it's probably fine to include it here as I think crucible should probably be
true by default.

Reply:
Good suggestion, thanks! Implemented in 5ebb5ed.